### PR TITLE
fix: generate command could fail when analysing relations on MongoDB when having one or more views

### PR DESCRIPTION
**Pull request by rap2hpoutre**
## Description

This fix is based on Sébastien's suggestion. We have to remove "system" collections every time we use `databaseConnection.collections()`.

## Refactor

👉  To do so, I had to refactor some code to avoid duplicating too much things. @ reviewer, we try to avoid mixing refactoring and a fix. For this particular case, fixing it without refactoring would have resulted maybe in more code than this actual commit. That's why I exceptionally decided to mix refactor and a fix. Feel free to criticize. Also, code climate ask me for **more** refactor in this PR, I will not follow its advice: since `detectReference ` uses sub-functions that are actually different in each files. It would have required too much work to refactor it.

## How to test

Sébastien gave me a sample database that is available in the click-up ticket. This database should fail during database analysis with master and work with this branch.

## Pull Request checklist:

- [x] Write an explicit title for the Pull Request, following [Conventional Commits specification](https://www.conventionalcommits.org/en/v1.0.0/#summary)
- [ ] Create automatic tests
- [ ] No automatic tests failures
- [x] Test manually the implemented changes
- [x] Review my own code (indentation, syntax, style, simplicity, readability)
- [x] Wonder if you can improve the existing code

### CODE REVIEW

File: utils/mongo-collections.js [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+function isSystemCollection(collection) {
+  const collectionName = getCollectionName(collection);
+  return collectionName && collectionName.startsWith('system.');
+}
+
+async function findCollectionMatchingSamples(databaseConnection, samples) {
+  return P.mapSeries(databaseConnection.collections(), async (collection) => {
+    if (isSystemCollection(collection)) return null;
```
Comment (rap2hpoutre) [inline utils/mongo-collections.js:17]: The actual fix is here.

**Comment by slimee** [comment]
A test is failing on my laptop `services > dumper > sequelize › should generate the env file` 
Run alone and globally it's the same.

DOCKER_DATABASE_URL is received but is not expected:
Screenshot 2020-11-04 at 16 51 19

Do you have the same?

**Review by slimee — CHANGES_REQUESTED** [review]
Since this PR introduces a new behavior, it should have a test covering the new case.

**Comment by rap2hpoutre** [comment]
Thanks for this review! I've just added a test, still I'm not satisfied by what I've done for two reasons:
 - I still can't reproduce the bug with another database than the one Sébastien gave me. Still, it seems to be the same as those that I manually create: I can't get what is different. So I had to create a database that seems identical with the one created.
 - The test needs some comments to understand the purpose: `system` tables are returned only with mongoDB >= 4. So the test can validate only this case.

However, I tested and the condition works: it ignore all `system.` tables when importing, that's what we want. BUT, I failed to prove via automated test that it does not work before the fix. 

TL;DR: I'm not OK with my test, but I failed to make a better one.

**Comment by rap2hpoutre** [comment]
@slimee I created a PR for this comment here: https://github.com/ForestAdmin/lumber/pull/485#issuecomment-721813548

**Review by slimee — COMMENTED** [review]
I am testing the PR, but I would like to show you there two comments before the end.

File: services/analyzer/mongo-collections-analyzer.js [hunk 2]
```diff
@@ -150,18 +154,11 @@ function analyzeMongoCollections(databaseConnection) {
...
-        if (!collectionName) {
-          return;
-        }
+      return P.each(collections, async (collection) => {
+        const collectionName = getCollectionName(collection);
 
-        // NOTICE: Ignore system collections.
-        if (collectionName.startsWith('system.')) {
+        // Ignore system collections and collection without a valid name.
+        if (!collectionName || isSystemCollection(collection)) {
```
[the remark this answers is not in the document]
Comment (rap2hpoutre) [reply services/analyzer/mongo-collections-analyzer.js:161]: I guess no. If it don't have a name, it is **not** considered as a system collection. So, it means `isSystemCollection` would return false.

```js
// Collection has no name
if (!collectionName || isSystemCollection(collection)) {
  // We are here
}

if (isSystemCollection(collection)) {
  // We are not here AAAARGH!
}
```

I guess the `isSystemCollection` behavior is OK: a collection that does not have a name **is actually not** a system collection.

File: test/services/database-analyzer/database-analyzer-mongo.test.js [hunk 3]
```diff
@@ -31,22 +32,27 @@ const expectedSubDocumentsAmbiguousIds = require('../../../test-expected/mongo/d
...
   return new MongoHelper(mongoUrl);
 }
 
-async function getAnalyzerOutputWithModel(mongoUrl, model) {
+async function getAnalyzerOutput(mongoUrl, callback) {
   const mongoHelper = await getMongoHelper(mongoUrl);
   const databaseConnection = await mongoHelper.connect();
   await mongoHelper.dropAllCollections();
-  await mongoHelper.given(model);
+  await callback(mongoHelper);
```
[the remark this answers is not in the document]
Comment (rap2hpoutre) [reply test/services/database-analyzer/database-analyzer-mongo.test.js:45]: I thought it would be more scalable to have this:

```
getAnalyzerOutputWithModel
getAnalyzerOutput
```

Over this:

```
getAnalyzerOutputWithModel // with an optional param callback or not
// Caveat: one more option in function, and a if, which increase cognitive complexity.
```

But it's an **opinion**, so I can change if you prefer.

**Review by slimee — APPROVED** [review]
🚀
